# <ins>`Data Science Python Practice Repository`</ins>

-- Ajit Mane (ajitmane36@gmail.com)

Welcome to the Data Science Python Practice Repository! This repository is designed to provide a platform for practicing Python programming in the field of data science. It offers a wide range of code snippets and exercises to help you enhance your Python skills, with a particular focus on data analysis, machine learning, and data visualization.

In [1]:
# Importing necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

Question 1:
Write a Python function that takes a list of numbers as input and returns a new list containing only the even numbers from the original list.

In [2]:
# Writting function for even numbers
def get_even_numbers(numbers):
    even_numbers = []
    for num in numbers:
        if num % 2 == 0:
            even_numbers.append(num)
    return even_numbers

# validating function
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
even_numbers = get_even_numbers(numbers)
print(even_numbers)

[2, 4, 6, 8, 10]


Question 2. Write a Python function that takes a list of strings as input and returns a new list containing only the strings that have a length greater than a given value.

In [3]:
# Defining function
def filter_strings_by_length(strings, length):
    filtered_strings = []
    for string in strings:
        if len(string) > length:
            filtered_strings.append(string)
    return filtered_strings

# Validating Function
strings = ["apple", "banana", "cherry", "kiwi", "mango"]
length = 5
filtered_strings = filter_strings_by_length(strings, length)
print(filtered_strings)

['banana', 'cherry']


Question 3:
Write a Python function that takes a string as input and counts the frequency of each character in the string. The function should return a dictionary where the keys represent the characters and the values represent their corresponding frequencies.

In [4]:
# Defining function
def char_frequency(string):
    '''Function takes string as input and return frequency of each character
    Args: string
    Return: frequency of each character'''
    frequency={}
    string=string.replace(' ','')
    unique_char=set(string)
    # Using for loop ti iterate over each character
    for char in unique_char:
        frequency[char]=string.count(char)
    return frequency

# Validation
string = "hello world"
character_frequency = char_frequency(string)
print(character_frequency)

{'r': 1, 'w': 1, 'o': 2, 'e': 1, 'l': 3, 'd': 1, 'h': 1}


Question 4:
Write a Python function that takes a list of integers as input and returns a new list containing only the prime numbers from the original list.

In [5]:
# Defining function
def prime_number(list_):
    '''Function takes list of numbers and return prime numbers
    Arg: List of number
    Return: return list of number'''
    from sympy import isprime
    prime_number_list=[]
    for num in list_:
        if isprime(num)==True:
            prime_number_list.append(num)
        else:
            pass
    return prime_number_list

# Validate  function
numbers = [2, 3, 4, 5, 6, 7, 8, 9, 10]
prime_numbers = prime_number(numbers)
print(prime_numbers)

[2, 3, 5, 7]


Question 5:
Write a Python function that takes a string as input and checks if it is a pangram or not. A pangram is a sentence that contains every letter of the alphabet at least once. The function should return True if the input string is a pangram, and False otherwise.

In [6]:
# defining function
def is_pangram(string):
    '''Fumction takes string as input asnd return string is pangram or not
    Arg: string as input
    Return: True or False'''
    alphabets=['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z']
    string=string.lower()
    for char in alphabets:
        if char not in string:
            return False
        else:
            return True
        
# Validate  function
sentence = "The quick brown fox jumps over the lazy dog"
is_pangram_result = is_pangram(sentence)
print(is_pangram_result)

True


Question 6. Write a function called reverse_words that takes a sentence as input and returns the sentence with the order of words reversed. Each word in the sentence should remain unchanged, but the order of the words should be reversed.

In [7]:
# Defining function
def reverse_words(string):
    '''Function takes string as input and return string as reverse by words in it
    Argument: String
    Return: Reverse string by words'''
    string=string.split()
    reverse_list=string[::-1]
    return ' '.join(reverse_list)

# Validation
sentence = "Hello world, how are you?"
reversed_sentence = reverse_words(sentence)
print(reversed_sentence)

you? are how world, Hello


Question 7. Write a function called remove_duplicates that takes a list as input and returns a new list with duplicate elements removed. The order of the elements should be maintained in the new list.

In [8]:
# Defining Function
def remove_duplicates(list_):
    '''Function takes list of integers as input and returns unique elements
    Args: List of integers
    Returns: unique elements'''
    list_=list(set(list_))
    return list_

# Validation
lst = [1, 2, 3, 2, 4, 1, 5]
result = remove_duplicates(lst)
print(result)

[1, 2, 3, 4, 5]


Question 8. Write a function called find_missing_number that takes a list of integers from 1 to N (inclusive), with one number missing, and returns the missing number.
For example, if the input list is [1, 2, 3, 5], the missing number is 4, so the function should return 4.

In [9]:
# Defining Function
def missing_number(list_):
    '''Function takes list as input and returns missing numbers
    Args: List of integers
    Returns: Missing number'''
    missing_number=[]
    max_num=max(list_)
    for i in range(0,max_num+1):
        if i not in list_:
            missing_number.append(i)
    return missing_number
    
# Validation
list_ = [1, 2, 3, 5]
missing_number = missing_number(list_)
print(missing_number)

[0, 4]


Question 9. Write a function called common_elements that takes two lists as input and returns a new list containing the common elements between the two input lists.
For example, if the input lists are [1, 2, 3, 4] and [3, 4, 5, 6], the common elements are [3, 4], so the function should return [3, 4].

In [10]:
# Defining function for finding common element
def common_elements(list1, list2):
    '''Function takes two lists of numbers and return common elements from these two list
    Args: Two list of integers
    Returns: Common elements from two lists'''
    common_elements=[]
    for num in list1:
        if num in list2:
            common_elements.append(num)
    return common_elements

# Validation of function
list1 = [1, 2, 3, 4]
list2 = [3, 4, 5, 6]
common_elements_list = common_elements(list1, list2)
print(common_elements_list)

[3, 4]


Question 10. Write a function called count_vowels that takes a string as input and returns the count of vowels (a, e, i, o, u) in the string.

In [11]:
# Count of Vowels in string
def count_vowels(string:str):
    '''Function takes string as input and returns frequency of vovels present in string
    Args: string
    Return: frequency of vovels from string'''
    vowels_frequency={}
    vovels='aeiou'
    string=string.replace(' ','').lower()
    for char in vovels:
        if char in string:
            vowels_frequency[char]=string.count(char)
    frequency=sum(vowels_frequency.values())
    return vowels_frequency, frequency

# Validation of function
# Validation
text = "Hello, how are you?"
vowel_count = count_vowels(text)
print(vowel_count)

({'a': 1, 'e': 2, 'o': 3, 'u': 1}, 7)


Question 11. Write a Python function that takes a list of integers as input and returns a new list that contains only the even numbers from the input list.

In [12]:
# Function to find even numbers from given list
def find_even_numbers(list_):
    '''Function takes list of integers as input and returns only even numbers from list
    Args: List of integers
    Returns: Even numbers from list'''
    even_numbers=[]
    for num in list_:
        if (num%2==0):
            even_numbers.append(num)
    return even_numbers

# Validation of function
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(find_even_numbers(numbers))

[2, 4, 6, 8, 10]


Question 12. Write a Python function that takes a string as input and returns the number of words in the string. Assume that words in the string are separated by spaces.

In [13]:
# Function for counting of words in string
def count_words(string):
    '''Function takes string as input and returns numbers of words
    Args: String
    Returns: Count of words in string'''
    count=0
    word_list=string.split() # We can also use count length of list to count words in list
    for word in word_list:
        if word in word_list: #Uning count method
            count+=1
    return count

# Validation
text = "This is a sample sentence."
print(count_words(text))

5


Question 13. Write a Python function that takes a list of strings as input and returns a new list containing the lengths of the strings in the input list.

In [14]:
# Defining function
def string_length(list_):
    '''Function takes list as input and returns length of each string
    Args: List of string
    Returns: Length of each string in list'''
    string_length=[]
    for string in list_:
        string_length.append(len(string))
    return string_length

# Validation of function
strings = ["Hello", "world", "Python", ""]
result=string_length(strings)
print(result)

[5, 5, 6, 0]


Question 14. Write a Python function that takes a list of integers as input and returns a new list containing only the odd numbers from the input list.

In [15]:
# Function to find odd numbers
def find_odd_numbers(list_):
    '''Function takes list of integers as input and returns odd numbers
    Args: List of numbers
    returns: Odd numbers from given list of numbers'''
    odd_numbers=[]
    for num in list_:
        if (num%2!=0):
            odd_numbers.append(num)
    return odd_numbers

# Validation
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(find_odd_numbers(numbers))

[1, 3, 5, 7, 9]


Question 15. Write a Python function called average_grade that takes a list of grades (integers) as input and returns the average grade. The average should be calculated as the sum of all grades divided by the total number of grades.
For example, given the list [80, 90, 75, 85, 95], the function should return 85.0, which is the average of the grades.

In [16]:
# Defining function for finding average
def average_grade(list_):
    '''Function takes list of numbers as input and returns average
    Args: List of integers
    Returns: Average of list'''
    average=sum(list_)/len(list_)
    return average

# Validation
grades = [80, 90, 75, 85, 91]
average = average_grade(grades)
print(average)

84.2


Question 16. Question: Write a function called count_words that takes a string as input and returns the count of words in the string. A word is defined as a sequence of non-whitespace characters separated by whitespace. The function should ignore leading and trailing whitespace and treat consecutive whitespace characters as a single delimiter.

In [17]:
# Defining function
def word_count(string):
    '''Function takes string as input and returns count of words
    Args: String
    Returns: Count of words'''
    word_list=string.strip().split()
    return len(word_list)

# Validation of function
# Validation
text = "   Hello,   how are   you?   "
word_count = word_count(text)
print(word_count)

4


Question 17. Write a Python function called find_max that takes a list of numbers as input and returns the maximum number from that list.

In [18]:
# Finding maximum number from list

def max_number(list_):
    '''Funtion takes list as input and returns maximum number from list
    Args: List of integers
    Returns: Maximum number'''
    return max(list_)

# Validation
list_ = [12, 45, 8, 27, 64, 5, 32]
print(max_number(list_))

64


Question 18: Write a Python function to check if a given string is a palindrome. Implement the function is_palindrome(s) that takes a string s as input and returns True if s is a palindrome, and False otherwise.

In [19]:
# Finding input is palindrome or not

def is_palindrome(input):
    '''Function check weather given input is palindrome or not
    Args: Input as string or integers
    Returns: input is palindrome or not'''
    input=str(input)
    if input==input[::-1]:
        return True
    else:
        return False
    
# Validation of function
list_=['radar', 'level', 'gain', 143, 282]
for i in list_:
    print(is_palindrome(i))

True
True
False
False
True


19. Python Program to Check if a Number is Odd or Even.

In [20]:
# Defining funcrion
def check_even_odd(number):
    '''Function takes interger as input and returns input is even or odd
    Args: interger
    Returns: Interger even or odd'''
    if (number%2==0):
        return 'Even'
    else:
        return 'Odd'

# Validation
print(check_even_odd(5))
print(check_even_odd(100))

Odd
Even


20. Write python program to find the area of a triangle when 3 side of trianle given.

In [21]:
#### Defining function
def Area_Of_triangle():
    '''Function takes length of three side as input and return area of trinagle
    Args: Three sides legnth as input
    Returns: Area of trianle'''
    import numpy as np
    a=float(input('length of side a:'))
    b=float(input('length of side b:'))
    c=float(input('length of side c:'))
    # Calculate semi perimeter
    s=(a+b+c)/2
    # Calculating area
    Area=np.sqrt(s*(s-a)*(s-b)*(s-c))
    return f'Area Of tringle is: {round(Area, 2)} unit'

# validation
print(Area_Of_triangle())

Area Of tringle is: 81.33 unit


21. Write python Program to Check Leap Year

In [22]:
def check_leap(year):
    '''Function takes year as input and returns whether the year is leap or not
    Args: year
    Returns: Leap or not'''
    if (year % 4 == 0):
        if (year % 100 == 0):
            if (year % 400 == 0):
                return f'Given year {year} is a leap year'
            else:
                return f'Given year {year} is not a leap year'
        else:
            return f'Given year {year} is a leap year'
    else:
        return f'Given year {year} is not a leap year'
        
# Validation
print(check_leap(1700))

Given year 1700 is not a leap year


22. Write Python Program to Find the Sum of Natural Numbers.

In [23]:
def find_sum_of_natural_numbers(n):
    '''Function to find the sum of natural numbers
    Args: n (integer) - the number of natural numbers to sum
    Returns: sum (integer) - the sum of natural numbers'''
    sum = 0
    for i in range(1, n+1):
        sum += i
    return sum

# Validation
n = int(input("Enter the number of natural numbers: "))
result = find_sum_of_natural_numbers(n)
print(f"The sum of the first {n} natural numbers is: {result}")


The sum of the first 10 natural numbers is: 55


23. Write a program in python for finding the last Fibonacci number for given number of elements.

In [24]:
# Writing function for counting last element of Fibonacci series
def Fibonacci_num(N):
    '''function takes number as input and print Fibonacci series upto given number
    Args: Integer
    Returns: Last element of Fibonacci series'''
    Fibonacci=[0,1]
    if len(Fibonacci)==N:
        return Fibonacci
    elif len(Fibonacci)>N:
        return [0]
    else:
        while len(Fibonacci)<N:
            next_element=Fibonacci[-1]+Fibonacci[-2]
            Fibonacci.append(next_element)
    return Fibonacci[-1]

# Validation
result=Fibonacci_num(10)
print(result)

34


24. write a python program to reverse words in a string.

In [25]:
# Defining function
def reverse_word(string):
    '''Function takes string as input and returns revese wrod with same squence
    Args: String
    Returns: string'''
    string_split=string.split()
    reverse_words=[]
    for word in string_split:
        words=word[::-1]
        reverse_words.append(words)
        
    return ' '.join(reverse_words)

# Validation of function
string='My name is mohit'
result=reverse_word(string)
print(result)

yM eman si tihom


25. Write Python program to convert Kilometres to Miles.

In [26]:
# Defining function
def km_to_miles_convertor(km):
    '''Function takes kilometres as integer and return conversion of input in miles
    Args: Input as kilometres
    Returns: Converts input in miles'''
    # 1 km is equal to 0.62137 miles
    miles=km*0.62137
    return f'{round(miles, 4)} miles'
    
#Validating
print(km_to_miles_convertor(5))

3.1068 miles


26. Write Python Program to check if a Number is Positive, Negative or Zero

In [27]:
# Defining function
def check_number(num):
    '''Function takes integers as input returns wether\
    input is positive, negative or zero.'''
    if num>0:
        return f"Given input {num} is positive"
    elif num<0:
        return f"Given input {num} is negative"
    else:
        return f"Given input {num} is zero"
    
# Validation
print(check_number(10))
print(check_number(0))
print(check_number(-5))

Given input 10 is positive
Given input 0 is zero
Given input -5 is negative


27. write python program to find area of rectangle

In [28]:
# defining function for finding area of rectanle
def area_of_rectangle(length, width):
    '''Functon takes length and width returns area of tringle
    Arg: Legth, width of tringle
    Returns: Area of tringle'''
    area=length*width
    return area

# Validation
print(area_of_rectangle(5, 3))

15


28. Compare two lists in Python.

In [29]:
def compare_list(list1, list2):
    '''Function takes two lists as input and returns whether the given two lists are equal or not.'''
    sorted_list1 = sorted(list1)
    sorted_list2 = sorted(list2)
    
    if len(sorted_list1) != len(sorted_list2):
        return 'Given two lists are not equal'
    
    for i in range(len(sorted_list1)):
        if sorted_list1[i] != sorted_list2[i]:
            return 'Given two lists are not equal'
    
    return 'Given two lists are equal'

# Validation
list1 = [10, 20, 30, 40, 50, 60]
list2 = [10, 20, 30, 50, 40, 70]
print(compare_list(list1, list2))

Given two lists are not equal


29. Print FizzBuzz if the integer is divided by 3 and 5, if divided by 3 and not 5, then print Fizz. If the number is divided by 5 but not by 3, then print Buzz, and if it is not divided by 3 and 5, then print this number.

In [30]:
# Defining function
def fizzBuzz(n):
    result = []
    for i in range(1, n+1):
        if (i%3==0) and (i%5==0):
            result.append('FizzBuzz')
        elif (i%3==0) and (i%5!=0):
            result.append('Fizz')
        elif (i%3!=0) and (i%5==0):
            result.append('Buzz')
        elif (i%3!=0) and (i%5!=0):
            result.append(i)
        
    for i in result:
        print(i)
        
# Validation
fizzBuzz(15)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz


30.  Python function that performs email validation.

In [31]:
# Importing necessary libraries
import re

# defining function
def validate_email(email):
    pattern = r'^[\w\.-]+@[\w\.-]+\.\w+$'  # Regular expression pattern for email validation

    if re.match(pattern, email):
        return True
    else:
        return False
    
# Validation of function
email_addresses = [
    "john.doe@example.com",
    "jane@example.co.uk",
    "alice123@gmail.com",
    "bob.smith123@company.net",
    "invalid.email",
    "noatsymbol.com",
    "spaces are not allowed@example.com",
    "special_chars!@gmail.com",
    "missingdomain@.com"
]

# validating
for email in email_addresses:
    print(validate_email(email))

True
True
True
True
False
False
False
False
False


31. Swap element from list from specific index.

In [32]:
# Defining function
def swap_elemnt_list(list_, swap_element,  N):
    '''Function takes list as input and swap element from list'''
    element_index=list_.index(swap_element)
    list_[element_index]=N
    return list_

# Validation
list_=[1,2,3,4]
result=swap_elemnt_list(list_, 3, 6)
print(result)

[1, 2, 6, 4]


32. Write a python program to find the minimum sum of factors of a number. 

In [33]:
# Defining function
def factor_sum(n):
    factors = []
    
    # Find all factors of the number
    for i in range(1, n + 1):
        if n % i == 0:
            factors.append(i)

    # Calculate the minimum sum of factors
    min_sum = sum(factors)

    return min_sum

# Get the number from the user
num = int(input("Enter a number: "))

# Calculate the minimum sum of factors
min_sum = factor_sum(num)

# Print the result
print("The minimum sum of factors for", num, "is", min_sum)

The minimum sum of factors for 9 is 13


33. Find second highest element in list.

In [34]:
# Defining function
def second_highest_element(list_):
    '''Function takes list_ as arguments and returns second highest element
    Args: list
    Returns: Second Highest element'''
    list_=sorted(list_)
    second_highest_element=list_[len(list_)-2]
    return second_highest_element

# Validation
list1=[14, 12, 25, 6, 8, 4, 0]
print(second_highest_element(list1))

14


34. Write a reverse string using for loop.

In [35]:
# Defining function
def reverse_string(string):
    '''Function takes string as input and returns reverse string
    Args: string
    Return: reverse input string'''
    word_list=[]
    for word in string.strip().split():
        word_list.append(word)
    return " ".join(word_list[::-1])

# Validation
string='I am Ajit Mane'
result=reverse_string(string)
print(result)

Mane Ajit am I


35. Write a program that would print all the odd numbers from 1 to 255.

In [36]:
# Defining function to print odd numbers between given range
def odd_number(number):
    '''Function takes interger as input and returns 
    odd numbers between range from 1 to input number
    Args: Integer
    Return: Odd numbers between 1 and input number'''
    odd_numbers=[]
    if number<1:
        print('Please, Enter Number Larger than 0')
    else:
        for i in range(1,number):
            if i%2!=0:
                odd_numbers.append(i)
    return odd_numbers

# Finding odd numbers
result=odd_number(20)
print(result)

[1, 3, 5, 7, 9, 11, 13, 15, 17, 19]


36. Write a program that would print all the numbers between given range.

In [37]:
# Defining function
def print_numbers(start, stop):
    '''Function print all numbers between given input range
    Args: start and stop integers
    Return: Numbers between given range'''
    return list(range(start+1, stop))

# Validation
result=print_numbers(1,12)
print(result)

[2, 3, 4, 5, 6, 7, 8, 9, 10, 11]


37. Write a program that takes an array, and prints the AVERAGE of the values in the array. For example for an array [2, 10, 3], your program should print an average of 5.

In [38]:
# Print average of array
def average_of_array(array):
    '''Function takes input an array and returns average of an array'''
    import numpy as np
    return sum(array)/len(array)
# Finding average
array=[2,10,3]
print(int(average_of_array(array)))

5


38. Write python program to convert list to flatten?

In [39]:
# defining function
def flatten_list(list_):
    '''Function takes list as input and return flatten list'''
    flatten_list=[]
    for item in list_:
        if isinstance(item, list):
            flatten_list.extend(item)
        else:
            flatten_list.append(item)
    return flatten_list
    
# Validation
nested_list = [1, [2, 3], [4, 5, 6], 7]
result=flatten_list(nested_list)
print(result)

[1, 2, 3, 4, 5, 6, 7]


39. Reverse string without indexing and data type changing?

In [40]:
# Reversing string without indexing and data type changing
def reverse_string(string):
    '''function takes string as input and return reverse string'''
    if len(string)<=1:
        return string
    return reverse_string(string[1:])+string[0]

# Test the function
string = "hello, Ajit Mane"
reversed_str = reverse_string(string)
print("Original String:", string)
print("Reversed String:", reversed_str)

Original String: hello, Ajit Mane
Reversed String: enaM tijA ,olleh


40. Return dictionary with number and its square till given range.

In [41]:
# Defining function
def square_dict(n):
    '''Function takes number as input and return the dictionary of number and its square'''
    dict_={num:num**2 for num in range(1,n+1)}
    return dict_
    
# Testing function
result=square_dict(12)
print(result)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144}


41. Return dictionary with number and its square till given range without using list comprehension.

In [42]:
# Defining function
def square_dict_(n):
    '''Function takes number as input and return the dictionary of number and its square'''
    keys=[]
    values=[]
    for  num in range(1,n+1):
        keys.append(num)
        values.append(num**2)
        
    dict_=dict(zip(keys, values))
    return dict_
    
# Testing function
result=square_dict_(10)
print(result)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


42. Write a program in Python to check whether an integer is Armstrong number or not.

In [43]:
# Defining function
def armstrong_num(int_):
    '''Function takes interger as input and return wether input is armstrong number or not
    Args: interger
    Returns: wether input is armstrong or not'''
    sum_=0
    string=str(int_)
    for i in range(0, len(string)):
        sum_=sum_+int(string[i])**len(string)
        
    if int_==sum_:
        return 'Given number is Armstrong'
    else:
        return 'Given number is not Armstrong'
      
# Testing function
numbers_list= [153, 370, 371, 407, 9474, 8208, 112]
for num in numbers_list:
    print(f'{num}')
    result=armstrong_num(num)
    print(result)
    print('--'*45)

153
Given number is Armstrong
------------------------------------------------------------------------------------------
370
Given number is Armstrong
------------------------------------------------------------------------------------------
371
Given number is Armstrong
------------------------------------------------------------------------------------------
407
Given number is Armstrong
------------------------------------------------------------------------------------------
9474
Given number is Armstrong
------------------------------------------------------------------------------------------
8208
Given number is Armstrong
------------------------------------------------------------------------------------------
112
Given number is not Armstrong
------------------------------------------------------------------------------------------


43. Write a program in Python to print the Fibonacci series using iterative method.

In [44]:
# Defining function
def fibonacci_series(num):
    '''Function takes number as input and returns
    Args: Number
    Returns: Fibonacci series'''
    fibonacci_series=[0,1]
    if len(fibonacci_series)==num:
        return fibonacci_series
    elif len(fibonacci_series)>num:
        return [0]
    else:
        while len(fibonacci_series)<num:
            next_element=fibonacci_series[-1]+fibonacci_series[-2]
            fibonacci_series.append(next_element)
    return fibonacci_series
# Testing Function
result=fibonacci_series(9)
print(result)

[0, 1, 1, 2, 3, 5, 8, 13, 21]


44. Check given string is palindrome or not.

In [45]:
# Defining function to check string is palindrome or not
def check_palindrome(string):
    '''Function takes string as input and return string is palindrome or not
    Args: String
    Return: String is palindrome or not'''
    if len(string)==1:
        return 'String Is Palindrome'
    elif string==string[::-1]:
        return 'String Is Palindrome'
    else:
        return 'String Is Not Palindrome'
    
# Validation
list_=['madam', 'Ajit']
for string in list_:
    result=check_palindrome(string)
    print(f'{string} - {result}')
    print('--'*45)

madam - String Is Palindrome
------------------------------------------------------------------------------------------
Ajit - String Is Not Palindrome
------------------------------------------------------------------------------------------


45. Print uncommon elements from list.

In [46]:
# Defining function
def find_uncommon_ele(list1, list2):
    '''Function takes two lists as input and returns uncommon elements from lists
    Args: Two Lists
    Return: Uncommon elements from given lists'''
    uncommon_elements=[]
    list1_unique_elemets=set(list1)
    list2_unique_elemets=set(list2)
    for element in list1_unique_elemets:
        if element not in list2:
            uncommon_elements.append(element)
    
    for ele in list2_unique_elemets:
        if ele not in list1:
            uncommon_elements.append(ele)
    return uncommon_elements

# Validation
test_list1 = [ [1, 2], [3, 4], [5, 6] ]
test_list2 = [ [3, 4], [5, 7], [1, 2] ]
result=find_uncommon_ele(str(test_list1), str(test_list2))
print(result)

['6', '7']


In [47]:
list1=[1,2,3,4,5,6,7]
list2=[4,5,6,7,8,9,10]
result=find_uncommon_ele(list1, list2)
print(result)

[1, 2, 3, 8, 9, 10]


46. The Geological Survey of India (GSI) agency wants to conduct a survey in Sahyadri ranges. They want to know the number of unique hills and valleys in the ranges. There is a series of mountains in Sahyadri ranges. They have done a survey and listed down the height of each mountain and now they want to count hills and valleys. Mountain will be counted as part of a hill if the closest non-equal height mountain is taller than both side mountains. Similarly, a mountain will be calculated as part of a valley if the closest non-equal height mountain is shorter than the both side mountains. Mountain will be part of the same hill or valley if the adjacent mountains have the same height. You need to write a code to help the GSI agency which will return the total of unique hills and valleys.

Sample Input:
nums= [2,4,1,1,6,5]
Sample Output:
3
Explanation:
For the input [2, 4, 1, 1, 6, 5], the code will return 3

In [48]:
# Defining Function
def count_hills_and_valleys(nums):
    n = len(nums)
    hills = valleys = 0

    for i in range(1, n - 1):
        if nums[i] != nums[i - 1] and nums[i] != nums[i + 1]:
            if nums[i - 1] > nums[i] and nums[i + 1] > nums[i]:
                hills += 1
            elif nums[i - 1] < nums[i] and nums[i + 1] < nums[i]:
                valleys += 1

    return hills + valleys

# Sample Input
nums = [2, 4, 1, 1, 6, 5]
# Sample Output
print(count_hills_and_valleys(nums))

2


47. Write a Python program to get the transpose of matrix and print the result in output.

In [49]:
# Defining function to transpose of matrix
import numpy as np
def transpose_matrix(matrix):
    '''Function takes matrix as input and returns its transpose
    Arg: Matrix
    Returns: Transpose'''
    output_matrix=np.transpose(matrix)
    return output_matrix

# Checking function
a=[[11,22,33],
   [44,55,66],
   [77,88,99]]

result=transpose_matrix(a)
print(result)

[[11 44 77]
 [22 55 88]
 [33 66 99]]


48. Write a function to add two matrix.

In [50]:
# Defining function
def matrix_sum(matrix1, matrix2):
    '''Function takes two matrix as input and return summation of it
    Args: Two matrix
    Return: Summation of two input matrix'''
    import numpy as np
    matrix1=np.array(matrix1)
    matrix2=np.array(matrix2)
    output_matrix=matrix1+matrix2
    return output_matrix

# Validation
matrix1=[[1,1,1],[2,2,2],[3,3,3]]
matrix2=[[4,4,4],[5,5,5],[6,6,6]]
matrix_sum(matrix1, matrix2)

array([[5, 5, 5],
       [7, 7, 7],
       [9, 9, 9]])

49. Write function to merge two Dictionaries in Python.

In [51]:
# Defining function to merge two dictionaries 
def merge_two_dictionaries(dict1, dict2):
    '''Function takes two dictionaries as input and returns merged dictionaries into one dict
    Args: dict1, dict2
    Return: Merged input dictionaries into single dictionary'''
    # Copying dict1 into merged_dict
    merged_dict=dict1.copy()
    
    # Iterate over each items in dict2 using for loop
    for keys, values in dict2.items():
        merged_dict[keys]=values
        
    return merged_dict

# Validating above function
dict1 = { 'Alexandra' : 27, 'Shelina Gomez' : 22, 'James' : 29, 'Peterson' : 30 }   
dict2 = { 'Jasmine' : 19,'Maria' : 26, 'Helena' : 30 }
merge_two_dictionaries(dict1, dict2)

{'Alexandra': 27,
 'Shelina Gomez': 22,
 'James': 29,
 'Peterson': 30,
 'Jasmine': 19,
 'Maria': 26,
 'Helena': 30}

50. Python Program to Find LCM of two numbers.

In [52]:
# Defining class
class lcm_calculator:
    def __init__(self, num1, num2):
        self.num1=num1
        self.num2=num2
    def lcm(self):
        import math
        lcm=math.lcm(self.num1, self.num2)
        return lcm
    
# Finding LCM
num1=int(input('Enter num1'))
num2=int(input('Enter num2'))
lcm_of_numbers=lcm_calculator(num1, num2)
# Printing result
print(f'The LCM of {num1} and {num2}  is {lcm_of_numbers.lcm()}')

The LCM of 8 and 6  is 24


51. Write a function to return sum of input array.

In [53]:
# Defining Function
def simpleArraySum(ar):
    # Write your code here
    '''Function takes array as input and returns sum of it
    Args: Integer Array
    Returns: Sum of input Array'''
    array_sum=sum(ar)
    return array_sum

# Testing function
simpleArraySum([14,16,22,98])

150

52. Convert a given string to int using a single line of code.

In [54]:
# Defining function
def string_to_int_func(string:str):
    '''Function takes string as input and change datatype to int'''
    output=int(string)
    output_dtype=type(output)
    return output, output_dtype

string='8767'
string_to_int_func(string)

(8767, int)

53. Write a code snippet to convert a string to a list.

In [55]:
# Defining function
def str_to_list_func(string:str):
    '''Function takes string as input and returns list  of string'''
    str_list=string.split(' ')
    return str_list

# Testing function
string='I will be there'
str_to_list_func(string)

['I', 'will', 'be', 'there']

54. Write a code snippet to reverse a string.

In [56]:
# Defining Function
def reversed_str(str:str):
    '''Fucntion takes string as input and return reverse string'''
    rev_str=str[::-1]
    print(f'Original String: {str}')
    print(f'Reversed strring: {rev_str}')

string='I love my India'
reversed_str(string)

Original String: I love my India
Reversed strring: aidnI ym evol I


55. Given an array of integers, calculate the ratios of its elements that are positive, negative, and zero. Print the decimal value of each fraction on a new line with  places after the decimal.

Note: This challenge introduces precision problems. The test cases are scaled to six decimal places, though answers with absolute error of up to  are acceptable.

In [57]:
# Defining function
def elements_ratio_in_array(array):
    '''Function takes an array as input and 
    returns the ratio of all positive, negative, and zero elements in the array'''
    
    # Check the length of the array
    array_length = len(array)
    
    if array_length == 0:
        print('Given input array is empty. Please input a valid array.')
        return
    
    # Count positive, negative, and zeroes in the given array   
    positive = 0
    negative = 0
    zeroes = 0
    
    for element in array:
        if element > 0:
            positive += 1
        elif element < 0:
            negative += 1
        else:
            zeroes += 1
            
    # Calculating ratios
    total = positive + negative + zeroes
    
    if total == 0:
        print("All elements in the array are zero.")
        return
    
    # Print the ratios with six decimal places
    print("{:.6f}".format(positive / total))
    print("{:.6f}".format(negative / total))
    print("{:.6f}".format(zeroes / total))

# Find the ratios of the array
arr = [-4, 3, -9, 0, 4, 1]
elements_ratio_in_array(arr)

0.500000
0.333333
0.166667


56. Write a Python program to count the number of characters (character frequency) in a string.

In [58]:
class ElementCount:
    def __init__(self, input_string):
        self.input_string = input_string

    def count_elements(self):
        output_dict = {}
        if len(self.input_string) == 0:
            return 'Please give a valid input string'

        for char in self.input_string:
            # Check if the character is already in the dictionary, if not, initialize it with 1
            if char in output_dict:
                output_dict[char] += 1
            else:
                output_dict[char] = 1

        return output_dict

# Finding frequency of given input string
string = 'Data Scientist'
counter = ElementCount(string)
result = counter.count_elements()

print("Character Frequency:")
for char, frequency in result.items():
    print(f"'{char}' : {frequency}")

Character Frequency:
'D' : 1
'a' : 2
't' : 3
' ' : 1
'S' : 1
'c' : 1
'i' : 2
'e' : 1
'n' : 1
's' : 1


57. Write the code to find the Fibonacci series upto the nth term.

In [59]:
# defing function
def fibonacci(N):
    '''Function takes integer as input and return its fibonacci series
    Args: Interger input
    Return:Fibonacci seris'''
    # initialize series
    fib_series=[0,1]
    
    if N==0:
        return [0]

    while len(fib_series)<N:
        next_num=fib_series[-1]+fib_series[-2]
        fib_series.append(next_num)
    return fib_series

# Validating function
result=fibonacci(10)
print(result)

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


58. Write code to Calculate frequency of characters in a string

In [60]:
# Defing function 
def frequency(input_string):
    '''Function takes input as string and return frequency of each character in string
    Args: string
    Return: Frequency dict'''
    frequency_dict = {}
    
    for char in input_string:
        frequency_dict[char]=frequency_dict.get(char, 0) + 1
        
    return frequency_dict

# Validating function
input_string= 'I am data science student'
frequency(input_string)

{'I': 1,
 ' ': 4,
 'a': 3,
 'm': 1,
 'd': 2,
 't': 3,
 's': 2,
 'c': 2,
 'i': 1,
 'e': 3,
 'n': 2,
 'u': 1}

59. 
Alice and Bob each created one problem for HackerRank. A reviewer rates the two challenges, awarding points on a scale from 1 to 100 for three categories: problem clarity, originality, and difficulty.

The rating for Alice's challenge is the triplet a = (a[0], a[1], a[2]), and the rating for Bob's challenge is the triplet b = (b[0], b[1], b[2]).

The task is to find their comparison points by comparing a[0] with b[0], a[1] with b[1], and a[2] with b[2].

If a[i] > b[i], then Alice is awarded 1 point.
If a[i] < b[i], then Bob is awarded 1 point.
If a[i] = b[i], then neither person receives a point.
Comparison points is the total points a person earned.

Given a and b, determine their respective comparison points.

Example

a = [1, 2, 3]
b = [3, 2, 1]
For elements *0*, Bob is awarded a point because a[0] .
For the equal elements a[1] and b[1], no points are earned.
Finally, for elements 2, a[2] > b[2] so Alice receives a point.
The return array is [1, 1] with Alice's score first and Bob's second

In [61]:
# Defining function
def compareTriplets(a, b):
    '''Function takes two arrays as input and returns an array representing Alice's and Bob's scores.
    Args: a, b - Arrays of equal length
    Return: Array with two elements - Alice's score and Bob's score'''
    
    a_points = 0
    b_points = 0
    
    for i in range(len(a)):
        if a[i] > b[i]:
            a_points += 1
        elif a[i] < b[i]:
            b_points += 1

    return [a_points, b_points]

# Validating function
a = [1, 2, 3]
b = [3, 2, 1]

result = compareTriplets(a, b)
print(result)


[1, 1]


60. Given a square matrix, calculate the absolute difference between the sums of its diagonals.
For example, the square matrix  is shown below:
[[1, 2, 3],
[4, 5, 6],
[9, 8, 9]] .
The left-to-right diagonal = 1+5+9=15. The right to left diagonal =3+5+9=17 . Their absolute difference is |15-17|=2 .

In [62]:
# Defining function
def absdiagonalDifference(arr):
    '''Function takes an array matrix as input and returns the absolute difference between its diagonal sums.
    Args: Array matrix
    Return: Absolute difference between its diagonal sums'''

    left_diagonal_element = []
    right_diagonal_element = []

    for i in range(len(arr)):
        left_diagonal_element.append(arr[i][i])
        right_diagonal_element.append(arr[i][len(arr) - 1 - i])

    sum_left_diagonal = sum(left_diagonal_element)
    sum_right_diagonal = sum(right_diagonal_element)

    absolute_diff = abs(sum_left_diagonal - sum_right_diagonal)

    return absolute_diff

# Testing function
arr = [[1, 2, 3], [4, 5, 6], [9, 8, 9]]
result = absdiagonalDifference(arr)
print(result)


2


61 . Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zer elements.
Note that you must do this in-place without making a copy of the array.
- Example 1:
Input: nums = [0,1,0,3,12]
Output: [1,3,12,0,0]
- Example 2:
Input: nums = [0]
Output: [0]

In [63]:
# Defining function which moves all zeros to the end of array
def move_zero_to_end(arr):
    ''''Function takes array as input and returns array with all zero element moves to end
    Args: Array
    Return: Array with zero element at end'''

    # Using for loop to iterate over each element in arr
    # Removing 0 from the array and add 0 to end of the array
    for i in arr:
        if i==0:
            arr.remove(0)
            arr.append(0)
            
    return arr

    # Validating Function

array=[1,0,4,6,8,0,10,0]
move_zero_to_end(array)

[1, 4, 6, 8, 10, 0, 0, 0]

62. Write an algorithm to determine if a number n is happy.
A happy number is a number defined by the following process:
- Starting with any positive integer, replace the number by the sum of the squares of its digits.
Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
- Those numbers for which this process ends in 1 are happy.
Return true if n is a happy number, and false if not.

In [64]:
# Defining function to check given number is happy number or not

def is_happy_number(number):
    '''function takes a number as input and returns whether it is a happy number or not
    Args: Number
    Return: Given number is a happy number or not'''
    seen = set()
    
    while number != 1 and number not in seen:
        seen.add(number)
        digits = [int(digit) for digit in str(number)]
        square_sum = sum(digit ** 2 for digit in digits)
        number = square_sum
    
    return number == 1

# Test the function
result = is_happy_number(19)
if result:
    print("It's a happy number.")
else:
    print("It's not a happy number.")

It's a happy number.


63. Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target. You may assume that each input would have exactly one solution, and you may not use the same element twice.
- You can return the answer in any order.
  - Example 1:
Input: nums = [2,7,11,15], target = 9
Output: [0,1] Explanation: Because nums [0] + nums [1] == 9, we return [0, 1].
  - Example 2:
Input: nums = [3,3], target = 6
Output: [0,1]

In [65]:
# Defining function
def find_element(list, target):
    ''''Function takes list as input and return element which sum equal to target'''
    for i in range(len(list)):
        for j in range(len(list)):
            if i !=j:
                if list[i]+list[j]== target:
                    return [list[i], list[j]]
                
# Testing function
list=[2, 11 ,15, 7]
target=9
find_element(list, target)

[2, 7]

64.  Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range that is missing from the array.
- Example 1:
Input: nums [3,0,1] =
Output: 2
Explanation: n = 3 since there are 3 numbers, so all numbers are in the range [0,3]. 2 is the missing number in the range since it does not appear in nums.
- Example 2:
Input: nums [0,1] =
Output: 2
Explanation: n = 2 since there are 2 numbers, so all numbers are in the range [0,2]. 2 is the missing number in the range since it does not appear in nums.
- Example 3:
Input: nums = [9,6,4,2,3,5,7,0,1]
Output: 8

Explanation: n = 9 since there are 9 numbers, so all numbers are in the range [0,9]. 8 is the missing number in the range since it does not appear in nums.

In [66]:
# Defining function
def find_missing_element(arr):
    '''Function takes array as input and return number in the range that is missing from array '''
    # Finding min and max from arra
    min_elm=min(arr)
    max_elm=max(arr)
    
    desired_array=range(min_elm,max_elm+1)
    
    missing_element=[]
    # Using for loop to check missing element
    for i in desired_array:
        if i not in arr:
            missing_element.append(i)
            
    return missing_element

# Testing function
list_input_array=[[3,0,1], [0,1], [9,6,4,2,3,5,7,0,1]]

for i in list_input_array:
    print(f'Input Array: {i}')
    print('Missing Element:', find_missing_element(i))
    print('--'*35)

Input Array: [3, 0, 1]
Missing Element: [2]
----------------------------------------------------------------------
Input Array: [0, 1]
Missing Element: []
----------------------------------------------------------------------
Input Array: [9, 6, 4, 2, 3, 5, 7, 0, 1]
Missing Element: [8]
----------------------------------------------------------------------


65. Given an integer array nums, return all the triplets [nums[i], nums[j], nums [k]] such that i != j, i != k, and j != k, and nums [i] + nums [j] + nums [k] == 0.

Notice that the solution set must not contain duplicate triplets.
- Example 1:
Input: nums= [-1,0,1,2,-1,-4]
Output: [[-1,-1, 2], [-1,0, 1]]
Explanation:
nums [0] + nums [1] + nums [2] = (-1) + 0 + 1 = 0.
nums [1] + nums [2] + nums [4] = 0 + 1 + (-1) = 0.
nums [0] + nums [3] + nums [4] (-1)+2+(-1) = 0.

The distinct triplets are [-1,0,1] and [-1,-1,2]. 

Notice that the order of the output and the order of the triplets does not matter.
- Example 2:
Input: nums = [0,1,1]
Output: []
Explanation: The only possible triplet does not sum up to 0.
- Example 3:
Input: nums = [0,0,0]
Output: [[0,0,0]]
Explanation: The only possible triplet sums up to 0.

In [67]:
# defining function
def find_sum_triplet(arr):
    '''Function takes array as input and return triplet which sum equal to zero'''
    output=[]
    for i in arr:
        for j in arr:
            for k in arr:
                if i!=j and i!=k and j!=k and i+j+k==0:
                    if [i, j, k] not in output:
                        output.append([i, j, k])
                
    return output

# Testing function
input_arrays=[[-1,0,1,2,-1,-4],[0,1,1],[0,0,0]]

for i in input_arrays:
    print(f'Input Array: {i}')
    print('Triplet Element:', find_sum_triplet(i))
    print('--'*35)

Input Array: [-1, 0, 1, 2, -1, -4]
Triplet Element: [[-1, 0, 1], [-1, 1, 0], [0, -1, 1], [0, 1, -1], [1, -1, 0], [1, 0, -1]]
----------------------------------------------------------------------
Input Array: [0, 1, 1]
Triplet Element: []
----------------------------------------------------------------------
Input Array: [0, 0, 0]
Triplet Element: []
----------------------------------------------------------------------


66. Get an column number as a input and convert it to excel column alphabet. Let the column number starts from 0 which represents A in excel column alphabet
- TestCase1:
SampleInput: 0
SampleOutput: A

- TestCase2:
SampleInput: 26
SampleOutput: AA

In [68]:
# Defning function
def convert_to_excel_column(num):
    '''Function take number as input and convert excel column alphabet'''
    column = ""
    while num >= 0:
        remainder = num % 26
        column = chr(65 + remainder) + column
        num = (num // 26) - 1
    return column

# Validating function
column_number = int(input("Enter a column number (starting from 0): "))
excel_column = convert_to_excel_column(column_number)
print("Excel Column:", excel_column)

Excel Column: BE


67. Given five positive integers, find the minimum and maximum values that can be calculated by summing exactly four of the five integers. Then print the respective minimum and maximum values as a single line of two space-separated long integers.

In [69]:
# Defining function
def find_min_max_sum_of_list(list_):
    # Sort the list in ascending order
    list_.sort()
    # Calculate the minimum and maximum values
    minimum_value = sum(list_[:-1])
    maximum_value = sum(list_[1:])

    # return the results as space-separated long integers
    return f"Minimum sum: {minimum_value}; Maximum sum: {maximum_value}"

# Testing function
integers = [5, 2, 9, 3, 8]
find_min_max_sum_of_list(integers)

'Minimum sum: 18; Maximum sum: 25'

68. Given an array of integers, calculate the ratios of its elements that are positive, negative, and zero. Print the decimal value of each fraction on a new line with 6 places after the decimal.

Note: This challenge introduces precision problems. The test cases are scaled to six decimal places, though answers with absolute error of up to 10^-4 are acceptable.

Example:
arr=[1,1,0,-1,-1]
There are n=5 elements, two positive, two negative and one zero. Their ratios are 2/5=0.400000 , 2/5=0.400000 and 1/5=0.200000. Results are printed as:

0.400000

0.400000

0.200000

In [70]:
# Defining function
def ratio_of_elements(array):
    '''Function takes array as input and return ratio of element as positive, negative and zero elements'''
    positive_elm=[]
    negative_elm=[]
    zero_elm=[]
    # Finding length of array
    n=len(array)
    
    # Finding positive, negative and zero element
    for i in array:
        if i<0:
            negative_elm.append(i)
        elif i==0:
            zero_elm.append(i)
        else:
            positive_elm.append(i)
            
    # Finding ratio
    positive_ratio = "{:.6f}".format(len(positive_elm) / n)
    negative_ratio = "{:.6f}".format(len(negative_elm) / n)
    zero_ratio = "{:.6f}".format(len(zero_elm) / n)
    
    # Printing result
    print(positive_ratio)
    print(negative_ratio)
    print(zero_ratio)

# Testing function
arr=[[1,1,0,-1,-1],[-4, 3, -9, 0, 4, 1]]

for i in arr:
    print(f'Input: {i}')
    print('Output')
    ratio_of_elements(i)
    print('--'*35)

Input: [1, 1, 0, -1, -1]
Output
0.400000
0.400000
0.200000
----------------------------------------------------------------------
Input: [-4, 3, -9, 0, 4, 1]
Output
0.500000
0.333333
0.166667
----------------------------------------------------------------------


69. You are in charge of the cake for a child's birthday. You have decided the cake will have one candle for each year of their total age. They will only be able to blow out the tallest of the candles. Count how many candles are tallest.

Example:
candles=[4,4,1,3]

The maximum height candles are  4 units high. There are 2 of them, so return 2.

In [ ]:
# Defning function
def birthdayCakeCandles(candles):
    '''Function takes as list and return number of highest candles'''
    
    highest_candles=max(candles)
    
    count=0
    
    for i in candles:
        if i==highest_candles:
            count=count+1
    return count

# Testing function
candles=[4,4,1,3]
birthdayCakeCandles(candles)

2

70. HackerRank University has the following grading policy:
- Every student receives a grade in the inclusive range from 0 to 100.
- Any grade less than 40 is a failing grade.

Sam is a professor at the university and likes to round each student's grade according to these rules:
- If the difference between the grade and the next multiple of 5 is less than 3, round grade up to the next multiple of 5.
- If the value of grade is less than 38, no rounding occurs as the result will still be a failing grade.

Given the initial value of grade for each of Sam's n students, write code to automate the rounding process.

In [1]:
# Defining function to rounding up grade of students
def round_grades(grades):
    '''function takes grade of student as input and return rounded grade'''
    rounded_grades = []
    for grade in grades:
        if grade < 38:
            rounded_grades.append(grade)
        else:
            next_multiple_of_5 = (grade // 5 + 1) * 5
            if next_multiple_of_5 - grade < 3:
                rounded_grades.append(next_multiple_of_5)
            else:
                rounded_grades.append(grade)
    return rounded_grades

# Testing function
grades = [73, 67, 38, 33, 91]
rounded = round_grades(grades)
print("Original Grades:", grades)
print("Rounded Grades:", rounded)

Original Grades: [73, 67, 38, 33, 91]
Rounded Grades: [75, 67, 40, 33, 91]


71. Convert temperature from Celsius to Fahrenheit.

In [3]:
# Defining function to convert temperature from celsius to fehrenheit
def temperature_cel_to_feht(x):
    '''Function takes temperature in celsius and returns in fehrenheit'''
    output=lambda x: (9/5*x)+32
    return round(output(x), 2)

# Verifing function
temp_cel=[25, 27, 30, 31, 34, 36]

for temp in temp_cel:
    print(temperature_cel_to_feht(temp))

77.0
80.6
86.0
87.8
93.2
96.8


72. Create a class to represent a basic calculator with add,  subtract, multiply, and divide methods.

In [10]:
# Defining class to doing basic calculator
class basic_calculator:
    def add(self, a,b):
        return a+b
    def substract(self, a,b):
        return a-b
    def multiply(self, a,b):
        return a*b
    def devide(self, a,b):
        if b==0:
            return 'Cannot devided by zero.'
        else:
            return round(a/b,2)

## Testing defined class
# Define variable
calculator=basic_calculator()
a = 5
b = 9
addition=calculator.add(5,9)
substract=calculator.substract(5,9)
multiply=calculator.multiply(5,9)
devide=calculator.devide(5,9)

# Print the result
print(f'Addition of {a} and {b}: {addition}')
print(f'Substraction of {a} and {b}: {substract}')
print(f'Multiplication of {a} and {b}: {multiply}')
print(f'Devidation of {a} and {b}: {devide}')

Addition of 5 and 9: 14
Substraction of 5 and 9: -4
Multiplication of 5 and 9: 45
Devidation of 5 and 9: 0.56


73. Write a Python class named Rectangle constructed from length and width and a method that will compute the area of a rectangle.

In [3]:
## Defining class

class Rectangle:
    def __init__(self, length, width):
        self.length = length
        self.width = width

    def area(self):
        return f'Area of given rectangle: {self.length * self.width} unit'

# Testing the class
rectangle_area = Rectangle(15, 10)
area = rectangle_area.area()
print(area)

Area of given rectangle: 150 unit


74. Given a list of integers, write a Python function to find all unique triplets in the list that sum up to a given target value?

In [2]:
# Defing functcion
from itertools import combinations

def find_triplets(nums, target):
    nums.sort()
    result = set() #For unique triplets
    
    for triplet in combinations(nums, 3):
        if sum(triplet) == target:
            result.add(tuple(sorted(triplet)))
    
    return list(result)

# Example
nums_list = [-1, 0, 1, 2, -1, -4]
target_sum = 0
result = find_triplets(nums_list, target_sum)
print(result)

[(-1, 0, 1), (-1, -1, 2)]


75. Write a python program to find the area of tringle.

In [13]:
## Defing function
def area_of_tringle():
    base=float(input('Length of base of tringle'))
    height=float(input('Height of tringle'))

    # Defing function to find area of the tringle
    def maths_computation(base, height):
        '''Function will take hieght and base as input and return area of the tringle'''
        area=0.5*base*height
        return area

    # Calculating area
    area_cal= maths_computation(base, height)
    
    # Printing final result
    print(f'Area of tringle having base:{base} and height:{height} is : {round(area_cal,2)}')


 # Finding area of the tringle for each base and height from above lists
area_of_tringle()

Length of base of tringle 12.5
Height of tringle 18.4


Area of tringle having base:12.5 and height:18.4 is : 115.0


76. Write a Python class that takes a list of integers as input and returns the sum of all the unique numbers in the list. If a number appears more than once in the list, it should only be added to the sum once.

In [13]:
# Defining class
class sum_unique_number:
    def __init__(self, num_list):
        self.num_list=num_list
    def sum_of_integers(self):
        list1 = set(self.num_list)
        sum_of_int = sum(list1)
        return sum_of_int

# Testing function
int_list=[4,5,6,7,8,9,10,4,5,6,7,8,9,10]
sum_of_list=sum_unique_number(int_list)
sum_of_list.sum_of_integers()

49

77.  Create a function that takes a number as input and prints
its multiplication table.

In [8]:
# Ddefining function
def multiplication_table(number):
    '''Function takes numberas input and returns it's multiplication table'''
    multiplication_table=[]
    multiplier=range(1,11)
    for i in multiplier:
        multiplication_table.append(number*i)
    return multiplication_table

# Testing function
print(multiplication_table(4))

[4, 8, 12, 16, 20, 24, 28, 32, 36, 40]


78. Create a program that takes a year as input and checks if 
it is a leap year or not. /

In [1]:
#Defining function
def is_leap_year(year):
    if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
        return True
    else:
        return False

# Taking input from the user
year = int(input("Enter a year: "))

if is_leap_year(year):
    print(year, "is a leap year.")
else:
    print(year, "is not a leap year.")


Enter a year:  2024


2024 is a leap year.


79. Given two lists of numbers, concatenate them into a 
single list.%

In [6]:
# Defining function
def concatenate_list(list1, list2):
    '''Function takes two list as input and return single concatenate list'''
    list1.extend(list2)
    return list1

# Testing function
list1=[1,2,3,4,5]
list2=[6,7,8,9,10]
concatenate_list(list1,list2)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

80. Create a program that finds the common elements between two lists and stores them in a new list.

In [11]:
# Defining function
def common_elements(list1, list2):
    '''Function that takes two list as input and return common elements'''
    common_elements=[]
    # Finding unique elements from given lists
    set1=set(list1)
    set2=set(list2)
    # Finding common elements
    set3=set1.intersection(set2)
    return set3

# Validating function
list1=[1,2,3,4,5]
list2=[4,5,6,7,8,9]
common_elements(list1, list2)

{4, 5}

81. Given a list of words, find the word with the maximum
length and its length.


In [9]:
# Defining function 
def max_length_word(list1):
    '''Function takes list as input and return word and its length which have maximum length'''
    length_of_words=[]
    for word in list1:
        length_of_words.append(len(word))
    max_word_index=length_of_words.index(max(length_of_words))
    max_len_word=list1[max_word_index]
    return f'Maximum length word: "{max_len_word}" and its Length: {len(max_len_word)}'

# Testing function
list=['Data', 'Machine', 'Health', 'Sunday', 'IBM']
max_length_word(list)

'Maximum length word: "Machine" and its Length: 7'

82. Create a class to represent a basic calculator with add, subtract, multiply, and divide methods.

In [2]:
# Defining class
class calculator:
    def __init__(self, a, b):
        self.a=a
        self.b=b
    def add(self):
        return self.a+self.b
    def substract(self):
        return self.a-self.b
    def multply(self):
        return self.a*self.b
    def devide(self):
        if self.b==0:
            print("Devident can't be zero")
        else:
            return round(self.a/self.b,2)

# Testing function
cal=calculator(10,2)
print(cal.add())
print(cal.substract())
print(cal.multply())
print(cal.devide())

12
8
20
5.0


83. Create a class to represent a book with attributes like title, author, and publication year.

In [11]:
# Defining class
class book:
    def __init__(self, title, author, publication_year):
        self.title=title
        self.author=author
        self.publication_year=publication_year
        
# Example
book1 = book("To Kill a Mockingbird", "Harper Lee", 1960)

print("Book 1:", book1.title, "by", book1.author, " Published in", book1.publication_year, '.')

Book 1: To Kill a Mockingbird by Harper Lee  Published in 1960 .


84. Define a class for a Circle with methods to calculate its area and circumference.

In [23]:
# Defining class
class circle:
    def __init__(self, radius):
        self.radius=radius
    def area(self):
        return f'Area of circle having radius:{self.radius} is {round(3.14*(self.radius)**2, 4)}'
    def circumference(self):
        return f'Circumference of circle having radius:{self.radius} is {round(2*3.14*self.radius, 4)}'
    
#Assignning object
my_circle=circle(5)
#Area of circle
print(my_circle.area())
#circumference of circle
print(my_circle.circumference())

Area of circle having radius:5 is 78.5
Circumference of circle having radius:5 is 31.4


85. Given a CSV file with employee details (name, position, salary), create a class to represent an Employee.

In [8]:
## Defining class

import csv

class Employee:
    def __init__(self, name, position, salary):
        self.name = name
        self.position = position
        self.salary = salary

    def display_details(self):
        print(f"Name: {self.name}, Position: {self.position}, Salary: {self.salary}")


# Reading the CSV file (example)

file_path = r"C:\Users\ajitm\Downloads\DS Projects\Temporary Files\employee_details.csv"

employees_list = []
with open(file_path, 'r') as file:
    csv_reader = csv.DictReader(file)
    for row in csv_reader:
        # Creating Employee objects from CSV data
        employee = Employee(row['Name'], row['Position'], float(row['Salary']))
        employees_list.append(employee)

# Displaying details of each employee
for employee in employees_list:
    employee.display_details()

Name: Ankita Chavan, Position: Manager (Data Analytics), Salary: 2400000.0
Name: Ajit Mane, Position: Data Scientist, Salary: 1200000.0
Name: Saurabh Kadam, Position: Softer Tester, Salary: 600000.0
Name: Ajit Patole, Position: Senior Data Scientist, Salary: 1800000.0
Name: Akash Gosavi, Position: Data Analyst, Salary: 700000.0
Name: Sonal Sharma, Position: Data Engineer, Salary: 1000000.0
Name: Kirthy Rathod, Position: Data Scientist Entern, Salary: 500000.0
Name: Swapnil Suryawanshi, Position: Data Analyst, Salary: 650000.0


86. Write a Python function that takes a list of numbers as input and returns the mean (average) of the numbers.

In [4]:
# Defining function
def find_mean(list_):
    '''given function takes list as input and return mean of numbers in list'''
    mean_=sum(list_)/len(list_)
    return round(mean_,2)

# Testing above function
list1=[9,8,7,6,5,4,3,8,25,49,5,98,122,65,556,125,2,1255,1456]
find_mean(list1)

200.42

87. Write a Python function that takes a list of numbers as input and returns the median of the numbers.

In [10]:
# Defning function to find median of list
def find_median(list_):
    '''Function takes a list as input and returns the median of numbers in the list'''
    import math
    
    list_.sort()
    n = len(list_)
    
    if n % 2 == 0:
        # If the list has an even number of elements, average the middle two
        middle_left = n // 2 - 1
        middle_right = n // 2
        median_value = (list_[middle_left] + list_[middle_right]) / 2
    else:
        # If the list has an odd number of elements, take the middle element
        middle_index = n // 2
        median_value = list_[middle_index]
    
    return median_value

# Validating the function
list1 = [1, 3, 6, 13, 27, 12]
result = find_median(list1)
print(result)

9.0


88. Write a Python program to swap two variables.

In [3]:
# Defing function 
def swap_numbers(num1, num2):
    '''Function takes two numbers as inpur and swap'''
    print(f'num1:{num1}, num2={num2}')
    num1, num2= num2, num1
    print(f'num1:{num1}, num2:{num2}')

# Validating function
num1=10
num2=20
swap_numbers(num1, num2)

num1:10, num2=20
num1:20, num2:10


89. Write a Python program to generate a random number from given range.

In [4]:
# Defining function to generate random number
def generate_randpm_number():
    '''function generate random number'''
    import numpy as np
    start_range=int(input('Enter staring range'))
    ending_range=int(input('Enter ending range'))
    number=np.random.randint(start_range, ending_range)
    return print(f'Random number in given  range : {number}')

# Generating random number
generate_randpm_number()

Random number in given  range : 130


90. Write a Python program to convert kilometers to miles.

In [13]:
# Defing class
class km_mile_convertor:
    def __init__(self, km):
        self.km=km
    def convert_to_mile(self):
        miles=self.km*0.621371
        return f'{self.km} km which is equal to {round(miles, 2)} miles'
    
# Tesing function
distance_convertor=km_mile_convertor(5)
distance_convertor.convert_to_mile()

'5 km which is equal to 3.11 miles'

91. Write a Production grade Python program to display calendar.

In [2]:
# Constructing python code to display clender
import calendar

def get_user_input():
    year = int(input("Enter the year (e.g., 2024): "))
    month = int(input("Enter the month (1-12): "))
    return year, month

def display_calendar(year, month):
    print(f"\nCalendar for {calendar.month_name[month]} {year}:\n")
    print(calendar.month(year, month)) # Print calender of given month in input year

def main():
    user_year, user_month = get_user_input()
    display_calendar(user_year, user_month)

if __name__ == "__main__":
    main()


Calendar for January 2024:

    January 2024
Mo Tu We Th Fr Sa Su
 1  2  3  4  5  6  7
 8  9 10 11 12 13 14
15 16 17 18 19 20 21
22 23 24 25 26 27 28
29 30 31



92. Write a Python function to display calendar.

In [1]:
def display_calender():
    '''Function takes year and month as input and return calender of given input
    Args: year and month number
    Return: calender of given month and year'''
    # Importing library
    import calendar
    
    # Taking input from user
    year = int(input('Enter year e.g 2024'))
    month = int(input('Enter month number (1-12)'))
    # Display calender
    print(f'Calender for {month} {year}:')
    print(calendar.month(year, month))
    
# Testing function
display_calender()

Calender for 2 2024:
   February 2024
Mo Tu We Th Fr Sa Su
          1  2  3  4
 5  6  7  8  9 10 11
12 13 14 15 16 17 18
19 20 21 22 23 24 25
26 27 28 29



93. Write a Python program to solve quadratic equation.

- The standard form of a quadratic equation is: 𝑎𝑥^2 + 𝑏𝑥 + 𝑐 = 0,
where a, b and c are real numbers and 𝑎 ≠ 0.

- The solutions of this quadratic equation is given by: (−𝑏 ± (𝑏^2 − 4𝑎𝑐 )^1/2)/(2𝑎)

In [2]:
### Defining production grade prgramm to solve quadratic equation

#Importing necessary libraries
import cmath

# Validating coefficients that a not 0
def validate_coefficients(a,b,c):
    '''Validating coeficient that a not 0'''
    if a==0:
        raise ValueError(f'Coefficient of "a" should not zero for quadratic equation')
    
# Solve quadratic equation
def solve_quadratic_equation(a,b,c):
    '''Function solve quadratic equation'''
    # Solving discriminant
    delta=cmath.sqrt(b**2-4*a*c)
    
    # Calculating solution using delta
    solution1=(-b+delta)/(2*a)
    solution2=(-b-delta)/(2*a)
    
    return solution1, solution2

def main():
    try:
        # Taking input from user
        a=float(input('Enter tha value of a:'))
        b=float(input('Enter tha value of b:'))
        c=float(input('Enter tha value of c:'))
        
        # Validate coefficient
        validate_coefficients(a,b,c)
        
        # Find solution
        solution=solve_quadratic_equation(a,b,c)
        
        if solution[0]==solution[1]:
            print(f'Given quadratic equation has one real solution : {solution[0]}')
        else:
            print(f'Given quadratic equation has two distinct solutions : {solution[0]} and {solution[1]}')
    
    except ValueError as ve:
        print(f'Error: {ve}')
        
    except Exception as ex:
        print(f'An unexpected error occured: {ex}')
        
        
if __name__== "__main__":
    main()

Given quadratic equation has two distinct solutions : (-0.3333333333333333+0.47140452079103173j) and (-0.3333333333333333-0.47140452079103173j)


94. Write a Python Program to Check Leap Year.

- ###### If the year is divisible by 4 AND not divisible by 100 OR If the year is divisible by 400. If either of these conditions is true indicating that the year is a leap year, otherwise, it returns False.

In [2]:
# Defining function to check leap year
def check_leap_year():
    '''Function will take year as input and return weather is leap or not
    Args: year as int
    Return: Weather year is leap or not'''
    year=int(input('Enter the year:'))
    if (year%4==0 and year%100!=0) or (year%400==0):
        print(f'{year} is leap year')
    else:
        print(f'{year} is not leap year')
        
def main():
    check_leap_year()
    
if __name__=="__main__":
    main()

2024 is leap year


95. Write a Python Program to Check Prime Number.
- ###### Prime Numbers: A prime number is a whole number that cannot be evenly divided by any other number except for 1 and itself. For example, 2, 3, 5, 7, 11, and 13 are prime numbers because they cannot be divided by any other positive integer except for 1 and their own value.

In [1]:
### Defining development grade problem

# Defining function to take input
def input_number():
    number=int(input('Enter whole number:'))
    if number<=1:
        print(f'Number less than or equal to 1 are not prime')
    else:
        return number
    
# Defining function to check number is prime or not
def check_prime(number):
    for i in range(2, number):
        if number%i==0:
            return False
    return True

def main():
    number=input_number()
    if check_prime(number):
        print(f'{number} is a prime number.')
    else:
        print(f'{number} is not a prime number.')

if __name__=="__main__":
    main()

13 is a prime number.


96. Write a Python Program to Print all Prime Numbers in an Interval of 1-10.

In [2]:
def get_input():
    '''Function takes range from to to'''
    from_ = int(input('Number From: '))
    to = int(input('Number To: '))
    return from_, to

def is_prime(from_, to):
    '''Function finds prime numbers in the given range'''
    for num in range(from_, to + 1):
        for i in range(2, num):
            if (num % i) == 0:
                break
        else:
            print(num)

def main():    # Taking input from user
    from_, to = get_input()
    # Finding prime numbers
    is_prime(from_, to)

if __name__ == "__main__":
    main()

5
7


97. Write a Python Program to Find the Factorial of a Number.

In [ ]:
## Defining function to find out factorial of given number

def find_factorial(num):
    ''''Function will take number as input and return factorial of it
    Arg: Number
    Return: Factorial'''
     

98. Write a Python Program to Display the multiplication Table.

99. Write a Python Program to Print the Fibonacci sequence.
 - Fibonacci sequence:
The Fibonacci sequence is a series of numbers where each number is the sum of the two
preceding ones, typically starting with 0 and 1. So, the sequence begins with 0 and 1, and
the next number is obtained by adding the previous two numbers. This pattern continues
indefinitely, generating a sequence that looks like this:
0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, and so on.

 - Mathematically, the Fibonacci sequence can be defined using the following recurrence
relation:
𝐹(0) = 0 𝐹(1) = 1 𝐹(𝑛) = 𝐹(𝑛 − 1) + 𝐹(𝑛 − 2)𝑓𝑜𝑟𝑛 > 1


100. Write a Python Program to Check Armstrong Number?
- Armstrong Number:
It is a number that is equal to the sum of its own digits, each raised to a power equal to the
number of digits in the number.
For example, let's consider the number 153:
  - It has three digits (1, 5, and 3).
  - If we calculate 1^3+5^3+3^3 , we get 1+125+27, which is equal to 153.

So, 153 is an Armstrong number because it equals the sum of its digits raised to the power
of the number of digits in the number.

- Another example is 9474:
  - It has four digits (9, 4, 7, and 4).
  - If we calculate 9^4+4^4+7^4+4^4 , we get 6561+256+2401+256 , which is also
equal to 9474.

Therefore, 9474 is an Armstrong number as well.